File: implementations/perl/string-search/v8-knuth-morris-pratt-d7l8/kmp.pl
use strict;
use warnings;

sub kmp_search {
  my ($text, $pattern) = @_;
  return () if length($pattern) == 0;
  my @lps = (0) x length($pattern);
  my $len = 0;
  my $i = 1;
  while ($i < length($pattern)) {
    if (substr($pattern, $i, 1) eq substr($pattern, $len, 1)) {
      $len++;
      $lps[$i++] = $len;
    } elsif ($len != 0) {
      $len = $lps[$len - 1];
    } else {
      $lps[$i++] = 0;
    }
  }
  my @matches;
  my ($ti, $pj) = (0, 0);
  while ($ti < length($text)) {
    if (substr($text, $ti, 1) eq substr($pattern, $pj, 1)) {
      $ti++;
      $pj++;
      if ($pj == length($pattern)) {
        push @matches, $ti - $pj;
        $pj = $lps[$pj - 1];
      }
    } elsif ($pj != 0) {
      $pj = $lps[$pj - 1];
    } else {
      $ti++;
    }
  }
  return @matches;
}

sub expect {
  my ($text, $pattern, $expected) = @_;
  my @actual = kmp_search($text, $pattern);
  die "$pattern failed" unless join(',', @actual) eq join(',', @$expected);
}

expect('ababcabcabababd', 'ababd', [10]);
expect('aaaaa', 'aa', [0, 1, 2, 3]);
expect('abcdef', 'gh', []);
expect('abc', 'abc', [0]);
print "perl kmp ok\n";

File: implementations/go/string-search/v8-knuth-morris-pratt-d7l8/kmp.go
package main

import (
  "fmt"
  "reflect"
)

func kmpSearch(text string, pattern string) []int {
  if len(pattern) == 0 { return []int{} }
  lps := make([]int, len(pattern))
  for i, length := 1, 0; i < len(pattern); {
    if pattern[i] == pattern[length] {
      length++
      lps[i] = length
      i++
    } else if length != 0 {
      length = lps[length-1]
    } else {
      lps[i] = 0
      i++
    }
  }
  matches := []int{}
  for i, j := 0, 0; i < len(text); {
    if text[i] == pattern[j] {
      i++
      j++
      if j == len(pattern) {
        matches = append(matches, i-j)
        j = lps[j-1]
      }
    } else if j != 0 {
      j = lps[j-1]
    } else {
      i++
    }
  }
  return matches
}

func expect(text string, pattern string, expected []int) {
  actual := kmpSearch(text, pattern)
  if !reflect.DeepEqual(actual, expected) {
    panic(fmt.Sprintf("%s: expected %v, saw %v", pattern, expected, actual))
  }
}

func main() {
  expect("ababcabcabababd", "ababd", []int{10})
  expect("aaaaa", "aa", []int{0, 1, 2, 3})
  expect("abcdef", "gh", []int{})
  expect("abc", "abc", []int{0})
  fmt.Println("go kmp ok")
}
